plugins/maven/src/index.ts

```diff
@@ -1,158 +1,391 @@
+
+  /**
+   * The password to use when deploying to the maven repository
+   *
+   * @deprecated since 9.38.0
+   **/
+  mavenPassword: t.string,
+});
+
+export type IMavenPluginOptions = t.TypeOf<typeof pluginOptions>;
+
+export interface IMavenProperties {
+  /** version **/
+  version?: string;
 
-  auto.logger.verbose.info(
-    "Maven: Got previous version from pom.xml",
-    previousVersion
-  );
+  /** git repository owner **/
+  owner?: string;
 
-  return previousVersion.replace("-SNAPSHOT", "");
+  /** git repository **/
+  repo?: string;
+
+  /** the author **/
+  developer?: IDeveloper;
 }
 
-/** Deploy project to maven */
+/** Deploy project to maven repository */
 export default class MavenPlugin implements IPlugin {
   /** The name of the plugin */
-  name = "maven";
+  readonly name = "maven";
+
+  /** The options of the plugin **/
+  private readonly options: Required<IMavenPluginOptions>;
+
+  /** cached properties **/
+  private properties: IMavenProperties = {};
+
+  /** should this release be a snapshot release **/
+  private snapshotRelease = false;
+
+  /** should pom.xml versions be handled with the "versions-maven-plugin" **/
+  private versionsMavenPlugin = false;
+
+  /** Initialize the plugin with its options **/
+  constructor(options: IMavenPluginOptions = {}) {
+    const {
+      MAVEN_COMMAND,
+      MAVEN_OPTIONS,
+      MAVEN_RELEASE_GOALS,
+      MAVEN_SETTINGS,
+      MAVEN_PASSWORD,
+      MAVEN_USERNAME,
+    } = process.env;
+    this.options = {
+      mavenCommand:
+        MAVEN_COMMAND ||
+        (options?.mavenCommand ? options.mavenCommand : "/usr/bin/mvn"),
+      mavenOptions: MAVEN_OPTIONS?.split(" ") || options.mavenOptions || [],
+      mavenReleaseGoals:
+        MAVEN_RELEASE_GOALS?.split(" ") ||
+        (options?.mavenReleaseGoals
+          ? options.mavenReleaseGoals
+          : ["deploy", "site-deploy"]),
+      mavenSettings: MAVEN_SETTINGS || options.mavenSettings || "",
+      printWidth: options.printWidth || 120,
+      tabWidth: options.tabWidth || 4,
+      mavenUsername: MAVEN_USERNAME || options.mavenUsername || "",
+      mavenPassword: MAVEN_PASSWORD || options.mavenPassword || "",
+    };
+  }
 
-  /** Tap into auto plugin points. */
-  apply(auto: Auto) {
-    auto.hooks.onCreateLogParse.tap(this.name, (logParse) => {
-      logParse.hooks.omitCommit.tap(this.name, (commit) => {
-        if (commit.subject.includes("[maven-release-plugin]")) {
-          return true;
-        }
-      });
-    });
+  /** Update the version in the pom.xml file **/
+  private static async updatePomVersion(
+    content: string,
+    version: string,
+    options: IMavenPluginOptions
+  ): Promise<string> {
+    const dom = new jsdom.JSDOM(content, { contentType: "text/xml" });
+    const pomDom = dom.window.document;
+    const versionNode = pomDom.evaluate(
+      "/project/version",
+      pomDom.documentElement,
+      pomDom.createNSResolver(pomDom.documentElement),
+      9
```

Comment:
What does the `9` mean?

Reply:
Great question - should have added a comment. It's the magic number for `XPathResult.FIRST_ORDERED_NODE_TYPE;`.

That particular object is not implemented by `jsdom` (though strangely it certainly uses the value correctly), so even though it's defined in the `lib.dom.d.ts` file, there were runtime exceptions.